Add feature to export tracks

sakertooth opened the pull request:
The option to export a track was added to the cog wheel for anything but automation tracks. Users are presented with a file dialog for the output location and then the export dialog to specify the output configuration.

### Future
This PR previously was for exporting clips as well, but this would require some refactor to do appropriately. This feature reuses some of the functionality already in `RenderManager`, but the problem is that we will eventually have 3-4 different export modes to choose from, which means a possibility for a lot of conditional branching and complexity.

My idea for a refactor might be to have a general `ExportContext` object, which holds the list of tracks and the time range to export, as well as if the tracks should be consolidated into one file or rendered into multiple files. This general approach should cover all the export modes we need, which will help streamline the `RenderManager` code and reduce complexity. This is just my idea for now, its subject to change.

### Notes for Testers
If testing, test that:
* The file dialog shows as expected when exporting the project, individual tracks, or a single track
* Switching the file type in the file dialog correctly switches the extension in the file name if the file name hasn't been changed
* The extension in the selected file name in the file dialog correctly matches the chosen file format in the export dialog
* The export settings work as usual (loop markers, add one bar, bit depth, bit rate etc)
* The exported audio sounds as expected.

Fixes #6666.

Spacemagehq commented:
I found a couple bugs for this pr there is an audio bug when you go to the volume tab in the instruments when you turn the amt on and all the way to the right the bounced audio/exported audio will get loud for no reason then go to normal volume. The other bug is the word cancel is cut off on Windows 11.
![new pr screenshot error](https://github.com/user-attachments/assets/6af33f88-af2e-4f4c-bc43-464f3481fb73)
![the cancel cut off](https://github.com/user-attachments/assets/d55a105d-2bae-4e3b-95ec-e2ea12f6bb4e)

sakertooth commented:
> I found a couple bugs for this pr there is an audio bug when you go to the volume tab in the instruments when you turn the amt on and all the way to the right the bounced audio/exported audio will get loud for no reason then go to normal volume. The other bug is the word cancel is cut off on Windows 11.

Are you also sure these issues don't appear on the master branch?

Spacemagehq commented:
> > I found a couple bugs for this pr there is an audio bug when you go to the volume tab in the instruments when you turn the amt on and all the way to the right the bounced audio/exported audio will get loud for no reason then go to normal volume. The other bug is the word cancel is cut off on Windows 11.
> 
> Are you also sure these issues don't appear on the master branch?

i'm on Windows, it's a pain in the ass to compile anything from scratch or build the master.

sakertooth commented:
> i'm on Windows, it's a pain in the ass to compile anything from scratch or build the master.

Can't you simply install the nightly build?

szeli1 reviewed (approved):
Please also add #8086 `MidiClipView.h` `SampleClipView.h`. You can improve them with `ProjectRenderer::availableDescriptions()`
Please apply the changes according to your opinion.

sakertooth commented on src/gui/modals/ExportProjectDialog.cpp:
```diff
@@ -34,21 +34,20 @@
+	, m_exportLocation(exportLocation)
+	, m_exportTracks(exportStems)
+	, m_trackToBounce(trackToBounce)
+	, m_renderManager(nullptr)
```
I used factory functions instead, which are more safer here IMO since it provides better guarantees that the correct constructor with the right parameters are called.

sakertooth commented on src/gui/modals/ExportProjectDialog.cpp:
```diff
@@ -34,21 +34,20 @@
-	setWindowTitle( tr( "Export project to %1" ).arg(
-					QFileInfo( _file_name ).fileName() ) );
+	setWindowTitle(tr("Export to %1").arg(QFileInfo(exportLocation).fileName()));
 
```
> explicitly enables markers "ExportMode::Clip".

I went with this idea.

szeli1 commented on include/RenderManager.h:
```diff
@@ -67,8 +74,8 @@ private slots:
 	void restoreMutedState();
-
 	void render( QString outputPath );
+	void populateUnmutedTracks();
```
Maybe `captureUnmutedTracks()` could be a better name.

szeli1 commented on src/gui/clips/MidiClipView.cpp:
```diff
@@ -935,5 +939,20 @@ bool MidiClipView::destructiveSplitClip(const TimePos pos)
+	dialog.setFileMode(FileDialog::Directory);
+	dialog.setAcceptMode(FileDialog::AcceptSave);
+	dialog.setDirectory(ConfigManager::inst()->userProjectsDir());
+	dialog.setWindowTitle(tr("Select directory for writing the exported track..."));
```
It was suggested to me to use `VersionedSaveDialog` or a dialog that uses the native file dialog. I'm unsure `FileDialog` does this.

szeli1 commented on src/gui/modals/ExportProjectDialog.cpp:
```diff
@@ -178,16 +176,39 @@ void ExportProjectDialog::startExport()
+	if (m_mode == Mode::ExportClip)
+	{
+		Engine::getSong()->setExportLoop(true);
+		exportLoopCB->setDisabled(true);
```
```suggestion
		exportLoopCB->setDisabled(true);
		Engine::getSong()->setExportLoop(true);
```

szeli1 reviewed (commented):
I did not test this PR. I miss the "drag n drop" functionality my PR provided. This PR does exporting clips better than my PR which is nice.

sakertooth committed:
Add export options in context menus

sakertooth committed:
Only implement exporting tracks

Exporting clips brings fourth a bit more variables and complexity that should be handled separately.

sakertooth committed:
Add file dialog for exporting tracks

sakertooth committed:
Add new export track mode

sakertooth committed:
Implement single track export

sakertooth committed:
Add space

sakertooth committed:
Merge remote-tracking branch 'upstream' into bounce-track

JohannesLorenz commented on src/gui/tracks/TrackOperationsWidget.cpp:
```diff
@@ -312,6 +343,14 @@ void TrackOperationsWidget::updateMenu()
 	}
+
+	// TODO: Allow for exporting tracks inside patterns
+	if (!dynamic_cast<AutomationTrackView*>(this)
```
Use `qobject_cast`, as it is usually more efficient? (same in following line)

sakertooth committed:
Merge remote-tracking branch 'upstream' into bounce-track

sakertooth committed:
Merge remote-tracking branch 'upstream' into bounce-track

sakertooth committed:
Merge remote-tracking branch 'upstream' into bounce-track

sakertooth committed:
Add asserts in RenderManager::renderTrack, iterate only through the track's container

sakertooth committed:
Move export mode back into the dialog

sakertooth committed:
Use setter for track instead of passing it in the constructor

sakertooth committed:
Fix formatting

sakertooth committed:
Refactor RenderManager to use a queue-based system

sakertooth committed:
Add imported exported track checkbox

sakertooth committed:
Import exported track as a new sample track

szeli1 commented on src/gui/modals/ExportProjectDialog.cpp:
```diff
@@ -202,9 +209,8 @@ ExportProjectDialog::ExportProjectDialog(const QString& path, RenderManager::Mod
 void ExportProjectDialog::onFileFormatChanged(int index)
 {
-	if (m_mode == RenderManager::Mode::ExportProject)
+	if (const auto fileInfo = QFileInfo{m_path}; !fileInfo.suffix().isEmpty())
```
Please separate these statements.

sakertooth replied:
Putting it in the if statement helps minimize scope, though I think here its mostly personal preference. I like it that way personally.

sakertooth replied:
Why not? I think minimizing scope is a good thing here nevertheless. I get that you might not like it because you rarely see it being done this way, but I think its a nice little trick. This was a feature added to if statements in C++17. I don't see any reason not to use this feature of the language. This isn't some implicit hack of any kind, its an actual feature of the language.

sakertooth replied:
[This](https://github.com/LMMS/lmms/wiki/Coding-Conventions) is our official coding conventions/style guidelines, and I don't see where something like that is mentioned.

szeli1 reviewed (changes requested):
I did not test this PR. I liked the overall changes!

sakertooth commented on include/RenderManager.h:
```diff
@@ -69,25 +60,29 @@ class RenderManager : public QObject
+	{
+		QString path;
+		std::vector<Track*> tracksToRender;
+	};
```
I still think `RenderManager` is unfit for exporting clips, even with this refactor. This refactor makes the plumbing a bit more mechanical, but that's it. The problem is that exporting a clip requires to mute all the other tracks and any overlapping clips. Ideally, `Song` would have direct support for exporting a track or clip like it does with `MidiClip` for playback in the Piano Roll, but maybe that's a long term solution that we will come back to.

Even more so, `Song` shouldn't be doing fifty different things at once, exporting being one of them.

szeli1 commented on src/core/RenderManager.cpp:
```diff
@@ -51,170 +50,146 @@ RenderManager::~RenderManager()
+
+	// TODO: Currently, only the song is exported (it will require changes and refactors in Song), but in the future we
+	// may want to generalize this function to work with any track container (e.g. the pattern store)
+	for (const auto& track : Engine::getSong()->tracks() | std::views::filter(&Track::isRenderable))
```
> Well, you knew what std::views::filter did right? Maybe not that unintuitive. 🤣

Yes I got the idea and I see that it is easy to understand what the code does. I don't like `std::views::filter` here because it can be replaced by an if statement without complicating the code. `std::views::filter` would make a lot of sense for passing a filtered array to a function (can it do that?), but since we directly control what will go in the loop I don't see a reason for this. Most c++ developers know if statements, and it is viewed as a standard method for filtering things, so why not use that?

szeli1 replied:
I agree that it is a neat, maybe novel way to filter things. Maybe it is true that it would be harder to forget filtering things. It is an unusual but memorable way to communicate to the programmer what the code does. What is the code actually doing? I have an assumption, but I don't know what does the "|" operator do. The statement also contains a function pointer. It is syntactically more complex than an if statement. This is an issue with my knowledge, I'm now looking into how `std::views::filter` works.

sakertooth replied:
I plan to hide it and centralize it in a function if that's better for you (which you simply cannot do with the if statement approach without making a whole new vector). Pretty sure I can but will have to try.

szeli1 commented on src/gui/tracks/TrackOperationsWidget.cpp:
```diff
@@ -274,8 +275,28 @@ void TrackOperationsWidget::exportTrack()
+		trackContainerView->moveTrackView(sampleTrackView, indexOfTrack + 1);
+		m_trackView->getTrack()->setMuted(true);
+		sampleClip->setSampleFile(exportPath);
+	});
```
Okay, keep it. It makes sense this way.

sakertooth committed:
Use factory methods for export dialog modes

sakertooth committed:
Add assert to ensure we are in single track mode

sakertooth committed:
Import exported track only if the dialog was accepted

sakertooth committed:
Remove use of std::views::filter, use if statement

JohannesLorenz commented on include/RenderManager.h:
```diff
@@ -40,15 +41,17 @@ class RenderManager : public QObject
-	void renderTracks();
+	void renderTracks(const QString& outputPath);
+
+	/// Export a a track into a single file
```
```suggestion
	/// Export a track into a single file
```

JohannesLorenz commented on include/TrackContainerView.h:
```diff
@@ -153,7 +153,10 @@ class TrackContainerView : public QWidget, public ModelView,
 	void realignTracks();
-	lmms::gui::TrackView * createTrackView( lmms::Track * _t );
+
+	//! Creates a track view for @a t if one doesn't exist already.
```
```suggestion
	//! Creates a track view for @a _t if one doesn't exist already.
```

sakertooth replied:
Why this change? What does the underscore do?

JohannesLorenz replied:
The parameter's name is `_t`, but you reference `t` in Doxygen? Might be even better to just fix the parameter name 😆

JohannesLorenz commented on include/TrackContainerView.h:
```diff
@@ -153,7 +153,10 @@ class TrackContainerView : public QWidget, public ModelView,
-	lmms::gui::TrackView * createTrackView( lmms::Track * _t );
+
+	//! Creates a track view for @a t if one doesn't exist already.
+	//! Otherwise, the already existing track view for @a t is returned.
```
```suggestion
	//! Otherwise, the already existing track view for @a _t is returned.
```

JohannesLorenz commented on include/RenderManager.h:
```diff
@@ -40,15 +41,17 @@ class RenderManager : public QObject
+	void renderTracks(const QString& outputPath);
+
+	/// Export a a track into a single file
+	void renderTrack(Track* track, const QString& outputPath);
```
Can we use a reference to `Track` here?

sakertooth replied:
Yes, but since we use pointers for these things, I figured its nicer to just call `renderTrack(track, ...)`, instead of `renderTrack(*track, ...)`.

JohannesLorenz commented on src/core/RenderManager.cpp:
```diff
@@ -50,146 +49,151 @@ RenderManager::~RenderManager()
-		// if we have not queued any tracks, renderNextTrack will just clean up
-		connect( m_activeRenderer.get(), SIGNAL(finished()),
-				this, SLOT(renderNextTrack()));
+	// TODO: We shouldn't need to allocate a new ProjectRenderer each time... we might also want to remove
```
This looks indeed weird to allocate here each time (and to reconnect the signals). I guess one could add something like `ProjectRenderer::changeOutputFilenameOfFileDev`, which changes the ProjectRenderer's output filename in case it's not renderering currently. Though, I wonder if that's easy to change, because you will need to change AudioFileWave, AudioFileFlac etc...

Maybe, to keep PRs atomic, it's a fair job for a separate PR.

sakertooth replied:
Yeah, we will have to refactor this in another PR. Things are a bit confusing right now.

JohannesLorenz commented on src/core/RenderManager.cpp:
```diff
@@ -50,146 +49,151 @@ RenderManager::~RenderManager()
 
-	if (m_tracksToRender.empty())
+	// TODO: Currently, only the song is exported (it will require changes and refactors in Song), but in the future we
+	// may want to generalize this function to work with any track container (e.g. the pattern store)
```
It looks like the old implementation of `renderTracks` did also render PatternStore tracks, but your new one does not do this.

Is there a reason for this? Usually, we need a good reason to remove existing features, I guess.

sakertooth replied:
IIRC, rendering patterns was never actually a feature to begin with.

`Song::startExport` was never created in such a way to handle rendering patterns, and rendering patterns wasn't an actual feature, so it was kind of just iterating over the pattern tracks for no real reason (maybe for some kind of "completeness", but it wasn't really doing anything).

sakertooth replied:
I might implement exporting tracks inside patterns honestly. I just hope it isn't too complicated.
Edit: On second thought, I might not introduce more changes to not make the reviewing anymore complicated than it already is. I will probably do this in another PR.

JohannesLorenz replied:
> I might not introduce more changes to not make the reviewing anymore complicated than it already is. I will probably do this in another PR.

100% agree. Let's keep the PRs small and atomic.

JohannesLorenz commented on include/RenderManager.h:
```diff
@@ -57,23 +60,29 @@ class RenderManager : public QObject
-	void restoreMutedState();
+	struct RenderJob
+	{
+		QString path;
```
I assume you can make these members const, to make it more obvious those will not be changed while in the queue? Or let the queue have only const member of this struct?

sakertooth replied:
> Or let the queue have only const member of this struct?

Sure :+1:

sakertooth replied:
Hmm, if I make the members of `RenderJob` `const`, it might make the class a bit weird to work with, and for some reason, `std::queue` cant have have a const `T` (probably because they need to move).

I could use `const` member functions, but then that would mostly just be extra boilerplate. Storing `std::unique_ptr<const RenderJob>` seems a bit over the top as well.

JohannesLorenz replied:
> I could use const member functions, but then that would mostly just be extra boilerplate. Storing std::unique_ptr<const RenderJob> seems a bit over the top as well.

Seems indeed OK. No change is needed then, I'll resolve the comment.

JohannesLorenz commented on src/core/RenderManager.cpp:
```diff
@@ -50,146 +49,151 @@ RenderManager::~RenderManager()
-				this, SIGNAL(progressChanged(int)));
+		if (!track->isRenderable()) { continue; }
+		track->setMuted(
+			std::find(job.tracksToRender.begin(), job.tracksToRender.end(), track) == job.tracksToRender.end());
```
Minor comment: This is `O(n*n)` worst/average case runtime. Would it make sense to make an `unordered_set` copy before the for loop, and then do a more descriptive `track->setMuted(!renderSet.contains(track));` (assuming our CI accepts that already)?

sakertooth replied:
That's an interesting idea. I haven't used `unordered_set`, but I'll look into it. I probably have to change some of this code anyway.

sakertooth replied:
Because of the cache locality advantages with `std::vector`, I'm not too sure performance in particular is an issue here, especially given that most projects dont have a lot of tracks. I think `unordered_set` can make it more readable with the `contains` method, but might not matter as much if I structure the code differently.

Then again, this is mostly speculation.

JohannesLorenz replied:
My assumption is that the maximum of tracks can be ~100, but even then, 100*100 times a boolean query is not much. On the other hand, readability is always nice, but also not a killer argument.

Please do as you prefer.

sakertooth replied:
I used `std::ranges::find` to simplify the line a little bit in 91ca221f82c3129719cc80a753eaac72417299ac.

JohannesLorenz commented on include/RenderManager.h:
```diff
@@ -57,23 +60,29 @@ class RenderManager : public QObject
-	std::vector<Track*> m_unmuted;
+	std::queue<RenderJob> m_renderJobQueue;
+	std::unordered_map<Track*, bool> m_muteStates;
+	int m_totalRenderJobs = 0;
```
To avoid some implicit casts:
```suggestion
	std::size_t m_totalRenderJobs = 0;
```

JohannesLorenz reviewed (commented):
Functional review finished. Most comments are editorial.

Some testing hints:

1. Test all 3 ways of exporting
2. Load/Save after importing an exported track
3. Press Ctrl + C (or kill LMMS) during export - is the project damaged on reopen?
4. Export projects with different types of tracks with different un/mute states - are the right tracks exported?

sakertooth committed:
Use size_t for m_totalRenderJobs variable

Co-authored-by: Johannes Lorenz <1042576+JohannesLorenz@users.noreply.github.com>

sakertooth committed:
Remove extra 'a' in docs

Co-authored-by: Johannes Lorenz <1042576+JohannesLorenz@users.noreply.github.com>

sakertooth committed:
Use qobject_cast where applicable

sakertooth committed:
Use std::ranges::find instead of std::find for simplicity and readability

sakertooth committed:
Remove underscore in _t parameter

sakertooth committed:
Merge remote-tracking branch 'upstream' into bounce-track

JohannesLorenz commented:
Rework is fine :+1:  The PR can be tested.
